pkg/templates/nstemplatetiers/nstemplatetier_generator_test.go

```diff
@@ -71,86 +142,141 @@ func TestCreateOrUpdateResources(t *testing.T) {
-						// as it seems like the real client/server don't not do it in the e2e tests
-						// see client.MockUpdate just underneath for the use-case
-						tier.ResourceVersion = ""
-					}
-					return err
-				}
-				return errors.Errorf("did not expect to create a resource of type %T", obj)
-			}
-			// check the 'generation' when updating the object, and increment its value
-			clt.MockUpdate = func(ctx context.Context, obj runtime.Object, opts ...client.UpdateOption) error {
-				if obj, ok := obj.(*toolchainv1alpha1.NSTemplateTier); ok {
-					if obj.ObjectMeta.ResourceVersion != "foo" {
-						// here, we expect that the caller will have set the ResourceVersion based on the existing object
-						return errors.Errorf("'ResourceVersion ' field must be specified during object update")
-					}
-					// increment even if the object did not change
-					existing := toolchainv1alpha1.NSTemplateTier{}
-					if err := clt.Get(ctx, types.NamespacedName{Namespace: obj.GetNamespace(), Name: obj.GetName()}, &existing); err != nil {
-						return err
-					}
-					obj.SetGeneration(existing.GetGeneration() + 1)
-				}
-				return clt.Client.Update(ctx, obj)
-			}
-			assets := nstemplatetiers.NewAssets(testnstemplatetiers.AssetNames, testnstemplatetiers.Asset)
+			clt.MockCreate = mockCreate(clt)
+			clt.MockUpdate = mockUpdate(clt)
 
 			// when
 			err := nstemplatetiers.CreateOrUpdateResources(s, clt, namespace, assets)
 			require.NoError(t, err)
-			// verify that 4 NSTemplateTier CRs were created: "advanced", "basic", "team", "nocluster"
+
+			// when calling CreateOrUpdateResources a second time
+			err = nstemplatetiers.CreateOrUpdateResources(s, clt, namespace, assets)
+
+			// then
+			require.NoError(t, err)
+			// verify that all TierTemplate CRs were updated
+			tierTmpls := toolchainv1alpha1.TierTemplateList{}
+			err = clt.List(context.TODO(), &tierTmpls, client.InNamespace(namespace))
+			require.NoError(t, err)
+			require.Len(t, tierTmpls.Items, len(testnstemplatetiers.AssetNames())-1) // exclude `metadata.yml` from the AssetNames, it does not result in a TemplateTier resource
+			for _, tierTmpl := range tierTmpls.Items {
+				assert.Equal(t, int64(1), tierTmpl.ObjectMeta.Generation) // unchanged
+			}
+
+			// verify that the 4 NStemplateTier CRs were updated
 			for _, tierName := range []string{"advanced", "basic", "team", "nocluster"} {
 				tier := toolchainv1alpha1.NSTemplateTier{}
 				err = clt.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: tierName}, &tier)
 				require.NoError(t, err)
-				// verify that the ResourceVersion was set by the server
-				assert.Equal(t, "foo", tier.ObjectMeta.ResourceVersion)
-				// verify that the generation was set to 1
-				assert.Equal(t, int64(1), tier.ObjectMeta.Generation)
-				// verify their namespace revisions
+				// verify that the generation was increased
+				assert.Equal(t, int64(2), tier.ObjectMeta.Generation)
```

Comment:
correct me if I'm wrong, the generation should be 1 because there was no change in any of the tiers - you called `nstemplatetiers.CreateOrUpdateResources` with the very same asset, so there should be anything updated/changed.

Reply:
addressed in 529d555